flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/DefaultScheduler.java

```diff
@@ -23,40 +23,93 @@
+import org.apache.flink.runtime.concurrent.ScheduledExecutor;
+import org.apache.flink.runtime.execution.ExecutionState;
+import org.apache.flink.runtime.executiongraph.ExecutionVertex;
+import org.apache.flink.runtime.executiongraph.failover.flip1.ExecutionFailureHandler;
+import org.apache.flink.runtime.executiongraph.failover.flip1.FailoverStrategy;
+import org.apache.flink.runtime.executiongraph.failover.flip1.FailureHandlingResult;
+import org.apache.flink.runtime.executiongraph.failover.flip1.RestartBackoffTimeStrategy;
 import org.apache.flink.runtime.executiongraph.restart.ThrowingRestartStrategy;
 import org.apache.flink.runtime.io.network.partition.PartitionTracker;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionID;
 import org.apache.flink.runtime.jobgraph.JobGraph;
+import org.apache.flink.runtime.jobmaster.LogicalSlot;
 import org.apache.flink.runtime.jobmaster.slotpool.SlotProvider;
 import org.apache.flink.runtime.metrics.groups.JobManagerJobMetricGroup;
 import org.apache.flink.runtime.rest.handler.legacy.backpressure.BackPressureStatsTracker;
+import org.apache.flink.runtime.scheduler.strategy.ExecutionVertexID;
+import org.apache.flink.runtime.scheduler.strategy.LazyFromSourcesSchedulingStrategy;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingStrategy;
+import org.apache.flink.runtime.scheduler.strategy.SchedulingStrategyFactory;
 import org.apache.flink.runtime.shuffle.ShuffleMaster;
+import org.apache.flink.runtime.taskmanager.TaskExecutionState;
 
 import org.slf4j.Logger;
 
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
 import java.util.concurrent.Executor;
 import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+import static org.apache.flink.util.Preconditions.checkState;
 
 /**
- * Stub implementation of the future default scheduler.
+ * The future default scheduler.
  */
-public class DefaultScheduler extends LegacyScheduler {
+public class DefaultScheduler extends SchedulerBase implements SchedulerOperations {
+
+	private final Logger log;
+
+	private final ClassLoader userCodeLoader;
+
+	private final ExecutionSlotAllocator executionSlotAllocator;
+
+	private final ExecutionFailureHandler executionFailureHandler;
+
+	private final ScheduledExecutor delayExecutor;
+
+	private final SchedulingStrategy schedulingStrategy;
+
+	private final ExecutionVertexVersioner executionVertexVersioner;
+
+	private final ExecutionVertexOperations executionVertexOperations;
 
 	public DefaultScheduler(
-			final Logger log,
```

Comment:
Should we have 2 indentations for params in function declaration, as suggested in the Flink [code style doc](https://flink.apache.org/contributing/code-style-and-quality-formatting.html)?

Reply:
We used to do this but in my opinion the code style doc isn't clear about it:

> Each new line should have one extra indentation (or two for a function declaration) relative to the line of the parent function name or the called entity

To be honest, I don't understand what it means. Clearly, it does not make sense to double indent the whole function declaration. Moreover, the example code which shows how to break function parameters, does not indent the parameters twice (I interpret 4 spaces as a single indentation):

```
public void func(
    int arg1,
    int arg2,
    ...) throws E1, E2, E3 {

}
```

Reply:
I think the benefit of double indentation is to avoid the param in function declaration mixed up with the function body which has single indentation. But as you did, the confusion can be reduced by adding an empty line between them.

I don't have much preference for this style actually, so it's fine to not change it if double indentation is not explicitly required.

Reply:
Since there is no clear guideline at the moment. I would leave it as is.

Reply:
Fine.